Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc

```diff
@@ -37,146 +36,182 @@ GEMGeometryBuilderFromCondDB::build(const std::shared_ptr<GEMGeometry>& theGeome
 	  
-	  superChamber->add( theGeometry->chamber( GEMDetId( detId.region(), detId.ring(), detId.station(), 1, detId.chamber(), 0 )));
-	  superChamber->add( theGeometry->chamber( GEMDetId( detId.region(), detId.ring(), detId.station(), 2, detId.chamber(), 0 )));
-	  ring->add( superChamber );
-	  theGeometry->add( superChamber );
-	  LogDebug("GEMGeometryBuilderFromDDD") << "Adding super chamber " << detId << " to ring: " 
-						<< "re " << re << " st " << st << " ri " << ri << std::endl;
+	  LogDebug("GEMGeometryBuilderFromDDD") << "Adding super chamber " << detId << " to ring: " << std::endl;
+	  ring->add(superChamber);
+	  theGeometry->add(superChamber);
+	} // end superChambers
+	
+	if (ring->nSuperChambers()){
+	  LogDebug("GEMGeometryBuilderFromDDD") << "Adding ring " <<  ri << " to station " << "re " << re << " st " << st << std::endl;
+	  station->add(ring);
+	  theGeometry->add(ring);
+	}
+	else {
+	  delete ring;
 	}
-	LogDebug("GEMGeometryBuilderFromDDD") << "Adding ring " <<  ri << " to station " << "re " << re << " st " << st << std::endl;
-	station->add( ring );
-	theGeometry->add( ring );
+      }// end ring
+      
+      if (station->nRings()){
+	LogDebug("GEMGeometryBuilderFromDDD") << "Adding station " << st << " to region " << re << std::endl;
+	region->add(station);
+	theGeometry->add(station);
       }
-      LogDebug("GEMGeometryBuilderFromDDD") << "Adding station " << st << " to region " << re << std::endl;
-      region->add( station );
-      theGeometry->add( station );
-    }
+      else {
+	delete station;
+      }
+    }// end station
+    
     LogDebug("GEMGeometryBuilderFromDDD") << "Adding region " << re << " to the geometry " << std::endl;
-    theGeometry->add( region );
+    theGeometry->add(region);
   }
+  
+}
+
+GEMSuperChamber* GEMGeometryBuilderFromCondDB::buildSuperChamber(const RecoIdealGeometry& rgeo, unsigned int gid, GEMDetId detId) const
+{
+  LogDebug("GEMGeometryBuilderFromCondDB") << "buildSuperChamber "<< detId <<std::endl;
+  
+  RCPBoundPlane surf(boundPlane(rgeo, gid, detId ));
+  
+  GEMSuperChamber* superChamber = new GEMSuperChamber(detId, surf);
+  return superChamber;  
+}
+
+GEMChamber* GEMGeometryBuilderFromCondDB::buildChamber(const RecoIdealGeometry& rgeo, unsigned int gid, GEMDetId detId) const
+{
+  LogDebug("GEMGeometryBuilderFromCondDB") << "buildChamber "<< detId <<std::endl;
+  
+  RCPBoundPlane surf(boundPlane(rgeo, gid, detId ));
+  
+  GEMChamber* chamber = new GEMChamber(detId, surf);
+  return chamber;
+}
+
+GEMEtaPartition* GEMGeometryBuilderFromCondDB::buildEtaPartition(const RecoIdealGeometry& rgeo, unsigned int gid, GEMDetId detId) const
+{
+  std::vector<std::string>::const_iterator strStart = rgeo.strStart( gid );
+  std::string name = *( strStart );
+  LogDebug("GEMGeometryBuilderFromCondDB") << "buildEtaPartition "<< name<<" "<< detId <<std::endl;
+  
+  std::vector<double>::const_iterator shapeStart = rgeo.shapeStart( gid );
+  float be = *(shapeStart+0)/cm;
```

Comment:
maybe there should be a size check here?

Reply:
all parameters for all detectors are stored in one vector<double> pars.
https://github.com/cms-sw/cmssw/blob/master/CondFormats/GeometryObjects/interface/RecoIdealGeometry.h#L118

Reply:
Yes, but you're still assuming that e.g. `(shapeStart+5)` is a valid reference...